### PR TITLE
[NR-57250] fix(tags): ignoring tags from different branches

### PR DESCRIPTION
**Pull request by paologallinaharbur**
Fix https://github.com/newrelic/release-toolkit/issues/109

The approach is quite simple, we check all tags and discard the ones not found in the list of commits since the root one (following the history).

```
rt validate-markdown --markdown "charts/newrelic-prometheus-agent"/CHANGELOG.md
rt generate-yaml --markdown "charts/newrelic-prometheus-agent"/CHANGELOG.md --included-dirs "charts/newrelic-prometheus-agent"  --tag-prefix newrelic-prometheus-agent-
INFO[0000] Ignoring newrelic-prometheus-agent-0.0.6 since it belongs to a different branch 
INFO[0000] Ignoring newrelic-prometheus-agent-0.0.7 since it belongs to a different branch 
INFO[0000] Ignoring newrelic-prometheus-agent-0.1.0 since it belongs to a different branch 
INFO[0000] Renovate source did not find any commit since "25153bf1101695b03f8a572b395f0070d39387fd" 
```
Screenshot 2022-10-11 at 16 56 06

### CODE REVIEW

File: src/git/tag.go [hunk 1]
```diff
@@ -42,8 +43,9 @@ var MatchAllTags = regexp.MustCompile("")
 
 func NewRepoTagsGetter(workDir string, opts ...TagOptionFunc) (*RepoTagsGetter, error) {
 	s := &RepoTagsGetter{
-		workDir: workDir,
-		match:   MatchAllTags,
+		workDir:       workDir,
+		match:         MatchAllTags,
+		commitsGetter: NewRepoCommitsGetter(workDir),
 	}
```
Comment (alvarocabanas) [inline src/git/tag.go:49]: Since You are composing the RepTagsGetter with a commitsGetter interface, then I would inject the NewRepoCommitsGetter from the call and change the signature to b eNewRepoTagsGetter(workDir string, commitsGetter CommitsGetter, opts ...TagOptionFunc).

**Review by alvarocabanas — COMMENTED** [review]
I left a small comment, apart from that, it looks good, it's the only approach we can take with go-git, what I would like is to see is the performance in a repo with many commits.

File: src/git/tag.go (hunk 1 above)
Comment (paologallinaharbur) [reply src/git/tag.go:49]: Makes sense!

File: src/app/generate/generate.go [hunk 2]
```diff
@@ -196,7 +196,8 @@ func tagVersionGetter(cCtx *cli.Context) (*git.TagsSource, error) {
 		tagOpts = append(tagOpts, git.TagsMatching("^"+matching))
 	}
 
-	src, err := git.NewRepoTagsGetter(cCtx.String(gitRootFlag), tagOpts...)
+	workDir := cCtx.String(gitRootFlag)
+	src, err := git.NewRepoTagsGetter(workDir, git.NewRepoCommitsGetter(workDir), tagOpts...)
```
Comment (roobre) [inline src/app/generate/generate.go:200]: I'm not a fan of putting a `RepoCommitsGetter` inside the `RepoTagsGetter`, I don' think the TagsGetter needs Commits to perform its basic functionality.

If I understood the code correctly, what we want to do here is to _filter_ tags after we have collected them, right? Meaning, we don't use any information from the tags that is not returned as `[]Tag` by `RepoTagsGetter.Tags`.

If this is the case I think that what we should do here is have a TagsFilterer that can work on what `RepoTagsGetter` returns.

```go
tagGetter := git.NewRepoTagsGetter(workDir, tagOpts...) // Same signature, object, and behavior as before this PR
tags := tagGetter.Tags()
//...
filterer := ReachableTagsFilterer(git.NewRepoCommitsGetter(workDir))
tags = filterer.Filter(tags)
```

File: src/git/tag.go [hunk 3]
```diff
@@ -94,3 +106,19 @@ func (s *RepoTagsGetter) Tags() ([]Tag, error) {
...
+		return nil, fmt.Errorf("getting git commits since empty tree: %w", err)
+	}
+
+	commitMap := map[string]bool{}
+	for _, c := range branchCommits {
+		commitMap[c.Hash] = true
+	}
+
+	return commitMap, nil
+}
```
Comment (roobre) [inline src/git/tag.go:124]: This dangling (i.e. not tied to an object) function is a very good pointer that we should have an object doing this work.

**Review by roobre — CHANGES_REQUESTED** [review]

File: src/app/generate/generate.go (hunk 2 above)
Comment (paologallinaharbur) [reply src/app/generate/generate.go:200]: I assumed that the basic functionality was retrieving the tags from a specific branch.

However, I can extract it, I just fear that in the future someone could use `.Tag`  without knowing that the filter is not applied by default

Comment (paologallinaharbur) [reply src/app/generate/generate.go:200]: would make sense to extract as well:
```
match         *regexp.Regexp
[...]
func TagsMatching(regex string) TagOptionFunc
```

Comment (paologallinaharbur) [reply src/app/generate/generate.go]: Now it follows the approach of the tags regex and it is "pluggable". 
If we want a separate all filtering then I would create something like `NewCommitFilter` for both the regex and the branch

File: src/git/tag.go [hunk 4]
```diff
@@ -19,21 +19,39 @@ type TagsGetter interface {
...
+	matchRegex   *regexp.Regexp
+	matchCommits map[string]bool
 }
 
 type TagOptionFunc func(s *RepoTagsGetter) error
 
-// TagsMatching returns an option that will make the getter to ignore tags that do not match regex.
-func TagsMatching(regex string) TagOptionFunc {
+// TagsMatchingRegex returns an option that will make the getter to ignore tags that do not match regex.
+func TagsMatchingRegex(regex string) TagOptionFunc {
```
Comment (roobre) [inline src/git/tag.go:30]: What is the motivation for this change? The name of the parameter here `regex string` should be already hinting what tags should match (a regex)

File: src/app/generate/generate.go (hunk 2 above)
Comment (roobre) [reply src/app/generate/generate.go]: > I see it as a similar functionality to: `func TagsMatching(regex string) TagOptionFunc`

This is a good point. The biggest difference I see is that matching a regex is a very easy thing to do (we don't require any dependency in the object), but filtering from a list of commits it's not, as it requires you to inject the CommitsGetter in the equation.

Adding `func TagsMatchingCommits(source CommitSource) TagOptionFunc` looks fine to me.

Extracting both regex filtering _and_ commit filtering out looks even better. In the end, we also have a tag replacement function, that is already extracted and performed by the `TagsSource`:
https://github.com/newrelic/release-toolkit/blob/e6cbc226f6c96c11d389232cfa7f92ce6774437e/src/git/tag_source.go#L35-L35

File: src/git/tag.go (hunk 4 above)
Comment (paologallinaharbur) [reply src/git/tag.go:30]: If you are in the file `A` and you see a function call:
```
a.TagsMatching("abc")
```

Are you sure what is matching? Is is a string, a prefix or a regex? That was the reasoning behind 😅

File: src/app/generate/generate.go (hunk 2 above)
Comment (paologallinaharbur) [reply src/app/generate/generate.go]: Ok, then we can leave it like this and wait for a major refactor when the v1 comes

File: src/app/nextversion/nextversion_test.go [hunk 5]
```diff
@@ -21,8 +21,8 @@ func TestNextVersion_Without_Repo(t *testing.T) {
 		globalargs string
 	}{
 		{
-			name:     "Overrides_Next",
-			args:     "-next v0.0.1",
+			name:     "Overrides_Next_And_Current",
+			args:     "-next v0.0.1 -current v1.2.3",
```
Comment (paologallinaharbur) [inline src/app/nextversion/nextversion_test.go:25]: Without a repo we cannot check current tags and commits

File: src/git/tag.go [hunk 6]
```diff
@@ -19,21 +19,39 @@ type TagsGetter interface {
...
 			return fmt.Errorf("compiling %q: %w", regex, err)
 		}
 
-		s.match = rgx
+		s.matchRegex = rgx
+		return nil
+	}
+}
+
+// TagsMatchingCommits returns an option that will make the getter to ignore tags that do not match commits.
```
Comment (roobre) [inline src/git/tag.go:42]:
```suggestion
// TagsMatchingCommits returns an option that will skip tags that do not point to a commit reachable from HEAD.
```

File: src/git/tag_source_test.go [hunk 7]
```diff
@@ -151,14 +214,16 @@ func TestRepoTagsSource_LastVersionHash(t *testing.T) {
 		{
 			name: "Matching_Leading_v",
 			tagOpts: []git.TagOptionFunc{
-				git.TagsMatching("^v"),
+				git.TagsMatchingRegex("^v"),
+				git.TagsMatchingCommits(commitsGetter),
```
Comment (roobre) [inline src/git/tag_source_test.go:218]: Why are we adding the `TagsMatchingCommits` opt on this test? It was passing before and we haven't changed the test requirements.

**Review by roobre — APPROVED** [review]
LGTM, left a small suggestion and a question regarding a test change. Good job!

File: src/git/tag_source_test.go (hunk 7 above)
Comment (paologallinaharbur) [reply src/git/tag_source_test.go:218]: I restored the old tests, there was no need for it
